src/libs/sys-util-rs/src/cgroup.rs

```diff
@@ -0,0 +1,690 @@
+}
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+lazy_static! {
+    /// Disable cgroup v1 subsystems.
+    pub static ref DISABLED_HIERARCHIES: Mutex<Vec<String>> = Mutex::new(Vec::new());
+}
+
+/// Update the disabled cgroup subsystems.
+pub fn update_disabled_cgroup_list(hierarchies: &[String]) {
+    let mut disabled_hierarchies = DISABLED_HIERARCHIES.lock().unwrap();
+    disabled_hierarchies.clear();
+    for hie in hierarchies {
+        disabled_hierarchies.push(hie.clone());
+    }
+    info!(
+        sl!(),
+        "disable cgroup list {:?} from {:?}", disabled_hierarchies, hierarchies
+    );
+}
+
+/// Filter out disabled cgroup subsystems.
+pub fn filter_disabled_cgroup(controllers: &mut Vec<Controllers>) {
+    let disabled_hierarchies = DISABLED_HIERARCHIES.lock().unwrap();
+    let mut blacklist_controllers: Vec<Controllers> = vec![];
+
+    for hierarchy in disabled_hierarchies.iter() {
+        match hierarchy.as_str() {
+            "cpu" => blacklist_controllers.push(cgroups::Controllers::Cpu),
+            "cpuset" => blacklist_controllers.push(cgroups::Controllers::CpuSet),
+            "cpuacct" => blacklist_controllers.push(cgroups::Controllers::CpuAcct),
+            "memory" => blacklist_controllers.push(cgroups::Controllers::Mem),
+            "blkio" => blacklist_controllers.push(cgroups::Controllers::BlkIo),
+            "perf_event" => blacklist_controllers.push(cgroups::Controllers::PerfEvent),
+            "devices" => blacklist_controllers.push(cgroups::Controllers::Devices),
+            "freezer" => blacklist_controllers.push(cgroups::Controllers::Freezer),
+            "hugetlb" => blacklist_controllers.push(cgroups::Controllers::HugeTlb),
+            "net_cls" => blacklist_controllers.push(cgroups::Controllers::NetCls),
+            "net_prio" => blacklist_controllers.push(cgroups::Controllers::NetPrio),
+            "pids" => blacklist_controllers.push(cgroups::Controllers::Pids),
+            "systemd" => blacklist_controllers.push(cgroups::Controllers::Systemd),
+            _ => (),
+        }
+    }
+    controllers.retain(|x| !blacklist_controllers.contains(x));
+}
+
+#[derive(Copy, Clone, Debug)]
+pub enum PidType {
+    /// Add pid to `tasks`
+    Tasks,
+    /// Add pid to `cgroup.procs`
+    CgroupProcs,
+}
+
+/// Get the singleton instance for cgroup v1 hierarchy object.
+pub fn get_cgroup_hierarchies() -> &'static cgroups::hierarchies::V1 {
+    static GLOBAL: Lazy<cgroups::hierarchies::V1> = Lazy::new(cgroups::hierarchies::V1::new);
+    GLOBAL.deref()
+}
+
+// Prepend a kata specific string to oci cgroup path to form a different cgroup path, thus cAdvisor
+// couldn't find kata containers cgroup path on host to prevent it from grabbing the stats data.
+const CGROUP_KATA_PREFIX: &str = "kata";
+
+/// Convert to a Kata specific cgroup path.
+pub fn gen_kata_cgroup_path(path: &str) -> String {
+    format!("/{}/{}", CGROUP_KATA_PREFIX, path.trim_start_matches('/'))
```

Comment:
It's usually better to do path building like this with `Path::join` rather than `format!`.

Reply:
Changed to
```
pub fn gen_kata_cgroup_path(path: &str) -> PathBuf {
    // Be careful to trim off the possible '/' prefix. Joining an absolute path to a `Path` object
    // will replace the old `Path` instead of concat.
    Path::new(CGROUP_KATA_PREFIX).join(path.trim_start_matches('/'))
}
```

Reply:
Looks good.